File: app/src/main/java/com/tt/handsomeman/ui/SignUp.java
package com.tt.handsomeman.ui;

import android.content.Intent;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.text.method.PasswordTransformationMethod;
import android.view.View;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.EditText;
import android.widget.ProgressBar;
import android.widget.Toast;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;

import com.tt.handsomeman.HandymanApp;
import com.tt.handsomeman.databinding.ActivitySignUpBinding;
import com.tt.handsomeman.model.SignUpFormState;
import com.tt.handsomeman.request.UserRegistration;
import com.tt.handsomeman.response.StandardResponse;
import com.tt.handsomeman.service.UserService;
import com.tt.handsomeman.util.SharedPreferencesUtils;
import com.tt.handsomeman.util.StatusCodeConstant;
import com.tt.handsomeman.util.StatusConstant;
import com.tt.handsomeman.viewmodel.SignUpViewModel;

import javax.inject.Inject;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class SignUp extends AppCompatActivity {

    @Inject
    SharedPreferencesUtils sharedPreferencesUtils;
    @Inject
    UserService userService;
    private SignUpViewModel signUpViewModel;
    private ActivitySignUpBinding binding;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivitySignUpBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        signUpViewModel = new ViewModelProvider(this).get(SignUpViewModel.class);

        final CheckBox cbPassword = binding.checkboxVisibleSignUpPassword;
        final CheckBox cbRePassword = binding.checkboxVisibleSignUpRePassword;
        final EditText edtPassword = binding.editTextSignUpPassword;
        final EditText edtRePassword = binding.editTextSignUpRePassword;
        final EditText edtName = binding.editTextSignUpYourName;
        final EditText edtMail = binding.editTextSignUpYourMail;
        final Button btnSignUp = binding.buttonSignUp;
        final ProgressBar pgSignUp = binding.progressBarSignUp;

        HandymanApp.getComponent().inject(this);

        binding.signUpBackButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                onBackPressed();
            }
        });

        observeSignUpState(edtPassword, edtRePassword, edtName, edtMail, btnSignUp);

        edtChangedListener(edtPassword, edtRePassword, edtName, edtMail);

        cbViewPassword(cbPassword, cbRePassword, edtPassword, edtRePassword);

        doSignUp(edtPassword, edtRePassword, edtName, edtMail, btnSignUp, pgSignUp);
    }

    private void doSignUp(EditText edtPassword,
                          EditText edtRePassword,
                          EditText edtName,
                          EditText edtMail,
                          Button btnSignUp,
                          ProgressBar pgSignUp) {
        String type = sharedPreferencesUtils.get("type", String.class);

        btnSignUp.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                pgSignUp.setVisibility(View.VISIBLE);
                btnSignUp.setEnabled(false);
                String name = edtName.getText().toString().trim();
                String mail = edtMail.getText().toString().trim();
                String password = edtPassword.getText().toString().trim();
                String rePassword = edtRePassword.getText().toString().trim();

                userService.doSignUp(type, new UserRegistration(name, mail, password, rePassword)).enqueue(new Callback<StandardResponse>() {
                    @Override
                    public void onResponse(Call<StandardResponse> call,
                                           Response<StandardResponse> response) {
                        if (response.body().getStatus().equals(StatusConstant.OK) && response.body().getStatusCode().equals(StatusCodeConstant.CREATED)) {
                            pgSignUp.setVisibility(View.INVISIBLE);
                            Toast.makeText(SignUp.this, response.body().getMessage(), Toast.LENGTH_LONG).show();
                            startActivity(new Intent(SignUp.this, Login.class));
                            finish();
                        } else {
                            pgSignUp.setVisibility(View.INVISIBLE);
                            btnSignUp.setEnabled(true);
                            Toast.makeText(SignUp.this, response.body().getMessage(), Toast.LENGTH_LONG).show();
                        }
                    }

                    @Override
                    public void onFailure(Call<StandardResponse> call,
                                          Throwable t) {
                        pgSignUp.setVisibility(View.INVISIBLE);
                        btnSignUp.setEnabled(true);
                        Toast.makeText(SignUp.this, t.getMessage(), Toast.LENGTH_LONG).show();
                    }
                });
            }
        });
    }

    private void cbViewPassword(CheckBox cbPassword,
                                CheckBox cbRePassword,
                                EditText edtPassword,
                                EditText edtRePassword) {
        cbPassword.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if (cbPassword.isChecked()) {
                    edtPassword.setTransformationMethod(null);
                } else {
                    edtPassword.setTransformationMethod(PasswordTransformationMethod.getInstance());
                }
                edtPassword.setSelection(edtPassword.length());
            }
        });

        cbRePassword.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if (cbRePassword.isChecked()) {
                    edtRePassword.setTransformationMethod(null);
                } else {
                    edtRePassword.setTransformationMethod(PasswordTransformationMethod.getInstance());
                }
                edtRePassword.setSelection(edtRePassword.length());
            }
        });
    }

    private void edtChangedListener(EditText edtPassword,
                                    EditText edtRePassword,
                                    EditText edtName,
                                    EditText edtMail) {
        TextWatcher afterTextChangedListener = new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s,
                                          int start,
                                          int count,
                                          int after) {
                // ignore
            }

            @Override
            public void onTextChanged(CharSequence s,
                                      int start,
                                      int before,
                                      int count) {
                // ignore
            }

            @Override
            public void afterTextChanged(Editable s) {
                signUpViewModel.signUpDateChanged(edtName.getText().toString().trim(), edtMail.getText().toString().trim(),
                        edtPassword.getText().toString().trim(), edtRePassword.getText().toString().trim());
            }
        };

        edtName.addTextChangedListener(afterTextChangedListener);
        edtMail.addTextChangedListener(afterTextChangedListener);
        edtPassword.addTextChangedListener(afterTextChangedListener);
        edtRePassword.addTextChangedListener(afterTextChangedListener);
    }

    private void observeSignUpState(EditText edtPassword,
                                    EditText edtRePassword,
                                    EditText edtName,
                                    EditText edtMail,
                                    Button btnSignUp) {
        signUpViewModel.getSignUpFormState().observe(this, new Observer<SignUpFormState>() {
            @Override
            public void onChanged(SignUpFormState signUpFormState) {
                if (signUpFormState == null) {
                    return;
                }
                btnSignUp.setEnabled(signUpFormState.isDataValid());
                if (signUpFormState.getNameError() != null) {
                    edtName.setError(getString(signUpFormState.getNameError()));
                }
                if (signUpFormState.getMailError() != null) {
                    edtMail.setError(getString(signUpFormState.getMailError()));
                }
                if (signUpFormState.getPasswordError() != null) {
                    edtPassword.setError(getString(signUpFormState.getPasswordError()));
                }
                if (signUpFormState.getRePasswordError() != null) {
                    edtRePassword.setError(getString(signUpFormState.getRePasswordError()));
                }
            }
        });
    }
}

File: app/src/main/java/com/tt/handsomeman/ui/SignUp.kt
package com.tt.handsomeman.ui

import android.content.Intent
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.text.method.PasswordTransformationMethod
import android.view.View
import android.widget.Button
import android.widget.CheckBox
import android.widget.EditText
import android.widget.ProgressBar
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders

import com.tt.handsomeman.HandymanApp
import com.tt.handsomeman.R
import com.tt.handsomeman.model.SignUpFormState
import com.tt.handsomeman.response.StandardResponse
import com.tt.handsomeman.service.SignUpService
import com.tt.handsomeman.util.SharedPreferencesUtils
import com.tt.handsomeman.util.StatusCodeConstant
import com.tt.handsomeman.util.StatusConstant
import com.tt.handsomeman.viewmodel.SignUpViewModel

import javax.inject.Inject

import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class SignUp : AppCompatActivity() {

    private var signUpViewModel: SignUpViewModel? = null

    @Inject
    internal var sharedPreferencesUtils: SharedPreferencesUtils? = null

    @Inject
    internal var signUpService: SignUpService? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_sign_up)

        signUpViewModel = ViewModelProviders.of(this).get(SignUpViewModel::class.java!!)

        val cbPassword = findViewById<CheckBox>(R.id.checkboxVisibleSignUpPassword)
        val cbRePassword = findViewById<CheckBox>(R.id.checkboxVisibleSignUpRePassword)
        val edtPassword = findViewById<EditText>(R.id.editTextSignUpPassword)
        val edtRePassword = findViewById<EditText>(R.id.editTextSignUpRePassword)
        val edtName = findViewById<EditText>(R.id.editTextSignUpYourName)
        val edtMail = findViewById<EditText>(R.id.editTextSignUpYourMail)
        val btnSignUp = findViewById<Button>(R.id.buttonSignUp)
        val pgSignUp = findViewById<ProgressBar>(R.id.progressBarSignUp)

        HandymanApp.component!!.inject(this)

        findViewById<View>(R.id.signUpBackButton).setOnClickListener { onBackPressed() }

        observeSignUpState(edtPassword, edtRePassword, edtName, edtMail, btnSignUp)

        edtChangedListener(edtPassword, edtRePassword, edtName, edtMail)

        cbViewPassword(cbPassword, cbRePassword, edtPassword, edtRePassword)

        doSignUp(edtPassword, edtRePassword, edtName, edtMail, btnSignUp, pgSignUp)
    }

    private fun doSignUp(edtPassword: EditText, edtRePassword: EditText, edtName: EditText, edtMail: EditText, btnSignUp: Button, pgSignUp: ProgressBar) {
        val type = sharedPreferencesUtils!!.get<String>("type", String::class.java)

        btnSignUp.setOnClickListener {
            pgSignUp.visibility = View.VISIBLE
            btnSignUp.isEnabled = false
            val name = edtName.text.toString()
            val mail = edtMail.text.toString()
            val password = edtPassword.text.toString()
            val rePassword = edtRePassword.text.toString()

            signUpService!!.doSignUp(type, name, mail, password, rePassword).enqueue(object : Callback<StandardResponse> {
                override fun onResponse(call: Call<StandardResponse>, response: Response<StandardResponse>) {
                    if (response.body()!!.status == StatusConstant.OK && response.body()!!.statusCode == StatusCodeConstant.CREATED) {
                        pgSignUp.visibility = View.GONE
                        Toast.makeText(this@SignUp, response.body()!!.message, Toast.LENGTH_LONG).show()
                        startActivity(Intent(this@SignUp, Login::class.java))
                        finish()
                    } else {
                        pgSignUp.visibility = View.GONE
                        btnSignUp.isEnabled = true
                        Toast.makeText(this@SignUp, response.body()!!.message, Toast.LENGTH_LONG).show()
                    }
                }

                override fun onFailure(call: Call<StandardResponse>, t: Throwable) {
                    pgSignUp.visibility = View.GONE
                    btnSignUp.isEnabled = true
                    Toast.makeText(this@SignUp, t.message, Toast.LENGTH_LONG).show()
                }
            })
        }
    }

    private fun cbViewPassword(cbPassword: CheckBox, cbRePassword: CheckBox, edtPassword: EditText, edtRePassword: EditText) {
        cbPassword.setOnClickListener {
            if (cbPassword.isChecked) {
                edtPassword.transformationMethod = null
            } else {
                edtPassword.transformationMethod = PasswordTransformationMethod()
            }
            edtPassword.setSelection(edtPassword.length())
        }

        cbRePassword.setOnClickListener {
            if (cbRePassword.isChecked) {
                edtRePassword.transformationMethod = null
            } else {
                edtRePassword.transformationMethod = PasswordTransformationMethod()
            }
            edtRePassword.setSelection(edtRePassword.length())
        }
    }

    private fun edtChangedListener(edtPassword: EditText, edtRePassword: EditText, edtName: EditText, edtMail: EditText) {
        val afterTextChangedListener = object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {
                // ignore
            }

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                // ignore
            }

            override fun afterTextChanged(s: Editable) {
                signUpViewModel!!.signUpDateChanged(edtName.text.toString(), edtMail.text.toString(),
                        edtPassword.text.toString(), edtRePassword.text.toString())
            }
        }

        edtName.addTextChangedListener(afterTextChangedListener)
        edtMail.addTextChangedListener(afterTextChangedListener)
        edtPassword.addTextChangedListener(afterTextChangedListener)
        edtRePassword.addTextChangedListener(afterTextChangedListener)
    }

    private fun observeSignUpState(edtPassword: EditText, edtRePassword: EditText, edtName: EditText, edtMail: EditText, btnSignUp: Button) {
        signUpViewModel!!.signUpFormState.observe(this, Observer { signUpFormState ->
            if (signUpFormState == null) {
                return@Observer
            }
            btnSignUp.isEnabled = signUpFormState.isDataValid
            if (signUpFormState.nameError != null) {
                edtName.error = getString(signUpFormState.nameError!!)
            }
            if (signUpFormState.mailError != null) {
                edtMail.error = getString(signUpFormState.mailError!!)
            }
            if (signUpFormState.passwordError != null) {
                edtPassword.error = getString(signUpFormState.passwordError!!)
            }
            if (signUpFormState.rePasswordError != null) {
                edtRePassword.error = getString(signUpFormState.rePasswordError!!)
            }
        })
    }
}
